network/src/testutils/test_framework.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    application::storage::PeerMetadataStorage,
+    peer_manager::{ConnectionRequestSender, PeerManagerRequestSender},
+    protocols::network::{NewNetworkEvents, NewNetworkSender},
+    testutils::test_node::{
+        ApplicationNetworkHandle, InboundNetworkHandle, NodeId, OutboundMessageReceiver, TestNode,
+    },
+};
+use channel::message_queues::QueueStyle;
+use diem_config::{
+    config::NodeConfig,
+    network_id::{NetworkId, PeerNetworkId},
+};
+use std::{collections::HashMap, hash::Hash, sync::Arc, vec::Vec};
+
+/// A trait describing a test framework for a specific application
+///
+/// This is essentially an abstract implementation, to get around how rust handles traits
+/// there are functions to get required variables in the implementation.
+///
+pub trait TestFramework<Node: TestNode + Sync> {
+    fn new(nodes: HashMap<NodeId, Node>) -> Self;
+
+    /// A constructor for `Node` specific to the application
+    fn build_node(node_id: NodeId, config: NodeConfig, peer_network_ids: &[PeerNetworkId]) -> Node;
+
+    fn nodes_mut(&mut self) -> &mut HashMap<NodeId, Node>;
+
+    /// In order to have separate tasks, we have to pull these out of the framework
+    fn take_node(&mut self, node_id: NodeId) -> Node {
```

Comment:
nit: I'd argue this should be a helper and not in the trait if it can be derived. Unless we plan on overriding it?